radicle-node/src/worker/fetch.rs

```diff
@@ -0,0 +1,330 @@
+mod refspecs;
+pub use refspecs::{AsRefspecs, Rad, Refspec};
+
+mod url;
+use url::LocalUrl;
+
+pub mod error;
+
+use std::collections::BTreeMap;
+use std::ops::Deref;
+
+use nonempty::NonEmpty;
+
+use radicle::crypto::{PublicKey, Unverified, Verified};
+use radicle::prelude::{Doc, Id};
+use radicle::storage::git::Repository;
+use radicle::storage::refs::{SignedRefs, IDENTITY_BRANCH};
+use radicle::storage::{Namespaces, RefUpdate, Remote, RemoteId};
+use radicle::storage::{ReadRepository, ReadStorage, WriteRepository, WriteStorage};
+use radicle::{git, storage, Storage};
+
+/// The initial staging area to perform a fetch from a remote.
+///
+/// The [`StagingRad::refpsecs`] generated are to fetch the `rad/id`
+/// and/or `rad/sigrefs` references from the remote end.
+///
+/// It is then expected to convert this into [`StagingRefs`] using
+/// [`StagingRad::into_staging_refs`] to continue the the rest of the
+/// references.
+pub struct StagingRad<'a> {
+    /// The inner [`Repository`] for staging fetches into.
+    pub(super) repo: StagedRepository,
+    /// The staging [`Storage`] for staging the `repo`.
+    staging: Storage,
+    /// The original [`Storage`] we are finalising changes into.
+    production: &'a Storage,
+    /// The `Namespaces` passed by the fetching caller.
+    namespaces: Namespaces,
+    _tmp: tempfile::TempDir,
+}
+
+/// Indicates whether the innner [`Repository`] is being cloned into
+/// or fetched into.
+pub enum StagedRepository {
+    Cloning(Repository),
+    Fetching(Repository),
+}
+
+impl Deref for StagedRepository {
+    type Target = Repository;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            StagedRepository::Cloning(repo) => repo,
+            StagedRepository::Fetching(repo) => repo,
+        }
+    }
+}
+
+/// The second staging area to perform a fetch from a remote.
```

Comment:
Area these actually different areas? I thought the refs shared the same staging area (repo)

Reply:
You're right :) It's the second _phase_ of fetching